main/src/ca/uwaterloo/flix/language/phase/typer/ConstraintGen.scala

```diff
@@ -574,13 +575,62 @@ object ConstraintGen {
         (resTpe, resEff)
 
       case Expr.StructNew(sym, fields, region, tvar, evar, loc) =>
-        throw new RuntimeException("joe tbd")
+        val (instantiatedFieldTpes, structTpe, regionVar) = instantiateStruct(sym, root.structs)
+        val visitedFields = fields.map { case (k, v) => visitExp(v) }
+        val (fieldTpes, fieldEffs) = visitedFields.unzip
+        // struct type
+        c.unifyType(tvar, structTpe, loc)
+        // field types
+        fields.zip(fieldTpes).foreach {
+          case ((fieldSym, expr), fieldTpe1) =>
+            instantiatedFieldTpes.get(fieldSym) match {
+              case None => () // Field does not exist. A resolution error will already have been generated
+              case Some(fieldTpe2) => c.unifyType(fieldTpe1, fieldTpe2, expr.loc)
+            }
+        }
+        // region type
+        val regionType = Type.mkRegion(regionVar, loc)
+        val (regionTpe, regionEff) = visitExp(region)
+        c.unifyType(regionType, regionTpe, region.loc)
+        // effect
+        c.unifyType(evar, Type.mkUnion(Type.mkUnion(fieldEffs, loc), regionEff, regionVar, loc), loc)
+
+        val resTpe = tvar
+        val resEff = evar
+        (resTpe, resEff)
 
-      case Expr.StructGet(sym, exp, name, tvar, evar, loc) =>
-        throw new RuntimeException("joe tbd")
+      case Expr.StructGet(sym, exp, field, tvar, evar, loc) =>
+        val (instantiatedFieldTpes, structTpe, regionVar) = instantiateStruct(sym, root.structs)
+        // struct type
+        val (tpe, eff) = visitExp(exp)
+        c.expectType(structTpe, tpe, exp.loc)
+        // field type
+        val fieldTpe = instantiatedFieldTpes(field)
+        c.unifyType(fieldTpe, tvar, loc)
+        // overall effect
+        c.unifyType(Type.mkUnion(eff, regionVar, loc), evar, exp.loc)
 
-      case Expr.StructPut(sym, exp1, name, exp2, tvar, evar, loc) =>
-        throw new RuntimeException("joe tbd")
+        val resTpe = tvar
+        val resEff = evar
+        (resTpe, resEff)
+
+      case Expr.StructPut(sym, exp1, field, exp2, tvar, evar, loc) =>
+        val (instantiatedFieldTpes, structTpe, regionVar) = instantiateStruct(sym, root.structs)
+        // lhs struct type
+        val (tpe1, eff1) = visitExp(exp1)
+        c.expectType(structTpe, tpe1, exp1.loc)
```

Comment:
As a style choice, I believe we tend to do all the recursive calls before we start expectingTypes